File: 2. 구현/자물쇠와 열쇠.js
"use strict"

function rotate(data) {
  const n = data.length
  const m = data[0].length
  const res = new Array(m).fill().map(() => new Array(n).fill(0))

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      res[j][i] = data[n - 1 - i][j]
    }
  }
  return res
}

function solution(key, lock) {
  const n = key.length
  const m = lock.length
  const hole = []

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < m; j++) {
      if (lock[i][j] === 0) hole.push([i, j])
    }
  }
  if (!hole.length) return true

  let repeat = 4
  while (repeat > 0) {
    const home = []
    key = rotate(key)
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        if (key[i][j] === 1) home.push([i, j])
      }
    }

    for (let [x, y] of home) {
      for (let [i, j] of hole) {
        const [dx, dy] = [-x + i, -y + j]
        const arrangedHome = home.map((el) => [el[0] + dx, el[1] + dy])

        let match = 0
        let cnt = 0
        for (let [ax, ay] of arrangedHome) {
          if (0 <= ax && ax < m && 0 <= ay && ay < m) {
            if (lock[ax][ay] === 0) {
              cnt += 1
              match += 1
            }
          } else cnt += 1
        }
        if (cnt === home.length && match === hole.length) return true
      }
    }
    repeat -= 1
  }
  return false
}


console.log(solution([
  [0, 0, 0],
  [1, 0, 0],
  [0, 1, 1]
], [
  [1, 1, 1],
  [1, 1, 0],
  [1, 0, 1]
]))

File: 2. 구현/자물쇠와 열쇠.py
def rotate(data):
    n = len(data)
    m = len(data[0])
    array = [[0] * n for _ in range(m)]
    for i in range(n):
        for j in range(m):
            array[j][i] = data[n - 1 - i][j]
    return array

def solution(key, lock):
  n = len(key)
  m = len(lock)
  hole = [[i,j] for i in range(m) for j in range(m) if lock[i][j] == 0]
  if (not len(hole)): return True
 
  for _ in range(4):
    key = rotate(key)
    home = [[i,j] for i in range(n) for j in range(n) if key[i][j]]
    
    for x, y in hole:
      for i, j in home:
        dx = -i + x
        dy = -j + y

        match = 0
        cnt = 0
        arrangedHome = [[i+dx,j+dy] for i, j in home]
        for ax, ay in arrangedHome:
          if (0<=ax<m and 0<=ay<m):
            if (lock[ax][ay] == 0):
              cnt += 1
              match += 1
          else:
            cnt += 1
        if match == len(hole) and cnt == len(home):
          return True
  return False

print(solution([[0, 0, 0], [1, 0, 0], [0, 1, 1]], [[1, 1, 1], [1, 1, 0], [1, 0, 1]]))